File: python/examples/stk_push.py
"""STK Push example using the M-Pesa Daraja SDK.

Run:

    export MPESA_CONSUMER_KEY=... MPESA_CONSUMER_SECRET=... \\
           MPESA_SHORTCODE=... MPESA_PASSKEY=...
    python stk_push.py
"""
from __future__ import annotations

import os
import sys

from mpesa import (
    Config,
    MpesaClient,
    MpesaError,
    STKPushRequest,
    TransactionType,
)


def main() -> None:
    # 1. Build config from environment variables.
    cfg = Config(
        consumer_key=os.environ["MPESA_CONSUMER_KEY"],
        consumer_secret=os.environ["MPESA_CONSUMER_SECRET"],
        shortcode=os.environ["MPESA_SHORTCODE"],
        passkey=os.environ["MPESA_PASSKEY"],
    )

    # 2. Create a client (shares the underlying requests.Session).
    client = MpesaClient(cfg)

    try:
        # 3. Send the STK Push prompt.
        resp = client.stk_push(STKPushRequest(
            transaction_type=TransactionType.CUSTOMER_PAY_BILL_ONLINE,
            amount=100,
            party_a="254712345678",
            phone_number="254712345678",
            call_back_url="https://example.com/callback",
            account_reference="Order 42",
            transaction_desc="test payment",
        ))

        # 4. Check acceptance (response_code "0" = Daraja accepted the request).
        #    Accepted is NOT paid — settle only via callback or STK Query.
        print(f"MerchantRequestID:  {resp.merchant_request_id}")
        print(f"CheckoutRequestID:  {resp.checkout_request_id}")
        print(f"ResponseCode:       {resp.response_code}")
        print(f"CustomerMessage:    {resp.customer_message}")

        if resp.is_accepted:
            print("\n✓ Push accepted — wait for callback or query with stk_query.")
        else:
            print(f"\n✗ Push rejected (code {resp.response_code}): "
                  f"{resp.response_description}")

    except MpesaError as exc:
        print(f"\n✗ Daraja error: {exc}", file=sys.stderr)
        if exc.error_code:
            print(f"  errorCode:    {exc.error_code}", file=sys.stderr)
        if exc.request_id:
            print(f"  requestId:    {exc.request_id}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: typescript/examples/stk_push.ts
/**
 * STK Push example using the M-Pesa Daraja SDK.
 *
 * Run:
 *
 *   export MPESA_CONSUMER_KEY=... MPESA_CONSUMER_SECRET=... \
 *          MPESA_SHORTCODE=... MPESA_PASSKEY=...
 *   npx tsx stk_push.ts
 */
import {
  Config,
  MpesaClient,
  MpesaError,
  TransactionType,
} from "@mpesa-sdk/core";

async function main(): Promise<void> {
  // 1. Build config from environment variables.
  const cfg = new Config({
    consumerKey:    process.env.MPESA_CONSUMER_KEY!,
    consumerSecret: process.env.MPESA_CONSUMER_SECRET!,
    shortcode:      process.env.MPESA_SHORTCODE!,
    passkey:        process.env.MPESA_PASSKEY!,
  });

  // 2. Create a concurrency-safe client (share across calls).
  const client = new MpesaClient({ config: cfg });

  try {
    // 3. Send the STK Push prompt.
    const resp = await client.stkPush({
      transactionType: TransactionType.BillPayGoods,
      amount: 100,
      partyA: "254712345678",
      phoneNumber: "254712345678",
      callBackURL: "https://example.com/callback",
      accountReference: "Order 42",
      transactionDesc: "test payment",
    });

    // 4. Check acceptance (ResponseCode "0" = Daraja accepted the request).
    //    Accepted is NOT paid — settle only via callback or STK Query.
    console.log(`MerchantRequestID:  ${resp.MerchantRequestID}`);
    console.log(`CheckoutRequestID:  ${resp.CheckoutRequestID}`);
    console.log(`ResponseCode:       ${resp.ResponseCode}`);
    console.log(`CustomerMessage:    ${resp.CustomerMessage}`);

    if (resp.ResponseCode === "0") {
      console.log("\n✓ Push accepted — wait for callback or query with stkQuery.");
    } else {
      console.log(
        `\n✗ Push rejected (code ${resp.ResponseCode}): ${resp.ResponseDescription}`,
      );
    }
  } catch (err) {
    if (err instanceof MpesaError) {
      console.error(`\n✗ Daraja error: ${err.message}`);
      if (err.errorCode) console.error(`  errorCode:    ${err.errorCode}`);
      if (err.requestId) console.error(`  requestId:    ${err.requestId}`);
    } else {
      console.error(`\n✗ Unexpected error: ${err}`);
    }
    process.exit(1);
  }
}

main();
